Fixes #3169: Item selection interaction text change and checkboxes should be disabled on selection

Akshatkamboj14 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fixes #3169
Fixes #3624

This PR is doing 2 things - 

1. It's updating the textView based on the selection of checkboxes by the user, If the user has not selected any checkbox then the Text will be `Please select all correct choices.`, If the user has selected less than maxSelection then the text will be `You may select more choices.` and if the user has selected maxSelection then the text will be `No more than $maxSelection choices may be selected.`

2. If the user has selected maximum checkboxes that can be selected, then the non selected checkboxes will be disbaled, and if the user unselects the checkbox or checkboxes, then all checkboxes will get enabled.

This is the link to the repository for the comparison of the StateFragmentTest.kt, to show that the same tests are failing on both branches(i.e. my feature branch and my develop branch)
https://github.com/Akshatkamboj14/Test-Comparison-For-StateFragmentTest.kt-oppia-android

<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

[issue.webm](https://user-images.githubusercontent.com/92685493/221786503-45d2c027-bedd-4cfb-92ce-3e351e81c34b.webm)

Akshatkamboj14 committed:
initial commit

Akshatkamboj14 committed:
corrected ktlint issues

Akshatkamboj14 committed:
corrected ktlint issues

Akshatkamboj14 committed:
correcting error

BenHenning commented:
As discussed during the meeting, there are some logic changes needed yet in the interaction view model. Please assign back once that's done @Akshatkamboj14.

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
loop for the enabled items

Akshatkamboj14 committed:
loop for the enabled items

BenHenning commented:
Will need to follow up with a code suggestion tomorrow.

BenHenning committed:
Merge branch 'develop' into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

BenHenning committed:
Ensure items disable when hitting max.

Ensure checkboxes have a disabled color.

BenHenning commented:
PTAL at https://github.com/Akshatkamboj14/oppia-android/pull/1 & the checklist for fixes and remaining things to do that I noticed as part of creating that PR. Merging it will automatically update your branch & show up on this PR.

Akshatkamboj14 committed:
Merge pull request #1 from BenHenning/suggested-fixes-for-4777

Ensure items disable when hitting max.

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'develop' into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection' of https://github.com/Akshatkamboj14/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
added instrumentation tests

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
corrected k-lint issues

Akshatkamboj14 committed:
added instrumentation tests

Akshatkamboj14 commented:
As @BenHenning Suggested, to attach the screenshot of the develop branch and my current feature branch to confirm that the failure is happening on both branches.
| My Feature Branch  | Develop Branch |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/92685493/212093610-c7d41687-cfa7-4505-bcb9-110dba009c93.png)  | ![image](https://user-images.githubusercontent.com/92685493/212132674-db2f9af0-4bc2-4fff-ba5e-19cfcee39d88.png)  |

BenHenning commented:
> As @BenHenning Suggested, to attach the screenshot of the develop branch and my current feature branch to confirm that the failure is happening on both branches.
> 
> My Feature Branch	Develop Branch
> ![image](https://user-images.githubusercontent.com/92685493/212093610-c7d41687-cfa7-4505-bcb9-110dba009c93.png)	![image](https://user-images.githubusercontent.com/92685493/212132674-db2f9af0-4bc2-4fff-ba5e-19cfcee39d88.png)

Is this the case when running all of the tests in StateFragmentTest on both develop and your branch? You should show the results for the full test suite in both to show that the same failures exist on both.

Also, this comparison belongs in the "For UI-specific PRs only" section of your PR description once it's completed.

Akshatkamboj14 commented:
> > As @BenHenning Suggested, to attach the screenshot of the develop branch and my current feature branch to confirm that the failure is happening on both branches.
> > My Feature Branch	Develop Branch
> > ![image](https://user-images.githubusercontent.com/92685493/212093610-c7d41687-cfa7-4505-bcb9-110dba009c93.png)	![image](https://user-images.githubusercontent.com/92685493/212132674-db2f9af0-4bc2-4fff-ba5e-19cfcee39d88.png)
> 
> Is this the case when running all of the tests in StateFragmentTest on both develop and your branch? You should show the results for the full test suite in both to show that the same failures exist on both.
> 
> Also, this comparison belongs in the "For UI-specific PRs only" section of your PR description once it's completed.

Okay Sure @BenHenning, I do that. There's one more thing, can you tell me whether the initial tests that I have pushed are correct or not and if I need more tests? I think only the `Add tests for the new AppCompatCheckBoxBindingAdapters in a similar way to other adapters tests.` this part is left. PTAL

Akshatkamboj14 committed:
merged-with-develop

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

BenHenning commented:
Sorry for the delay here @Akshatkamboj14. I think your tests seem reasonable, so I think you can go ahead and finish up the rest of the PR and send it into review if there are no other remaining questions.

Akshatkamboj14 committed:
added the previous enabled color

Akshatkamboj14 commented:
Hey @BenHenning, I have only one question Can you tell me how I can add tests for the binding adapters?

Akshatkamboj14 committed:
added AppCompatBindingAdapterTest

Akshatkamboj14 committed:
lint issue

Akshatkamboj14 committed:
syntax error

Akshatkamboj14 committed:
lint issue

Akshatkamboj14 committed:
lint issue

Akshatkamboj14 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+    activityRule.scenario.onActivity {
+      val appCompatCheckBox: AppCompatCheckBox = getAppCompatCheckBox(it)
+      setButtonTint(appCompatCheckBox, colorRgb)
+      assertThat((appCompatCheckBox.buttonTintList)?.defaultColor).isEqualTo((colorRgb))
```
Hey @BenHenning can you tell me what I am doing wrong on this?

BenHenning replied:
As discussed during meeting, I suggest looking through AOSP code search for these components to get an idea of reasons why null might be returned.

Akshatkamboj14 replied:
Okay Ben.

Akshatkamboj14 committed:
added AppCompatCheckBoxBindingAdaptersTestActivity into test_file_exemptions.textproto

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

Akshatkamboj14 committed:
corrected failing test

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -127,7 +127,7 @@ class AppCompatCheckBoxBindingAdaptersTest {
       val appCompatCheckBox: AppCompatCheckBox = getAppCompatCheckBox(it)
       setButtonTint(appCompatCheckBox, colorRgb)
-      assertThat((appCompatCheckBox.buttonTintList)?.defaultColor).isEqualTo((colorRgb))
+      assertThat((appCompatCheckBox.supportButtonTintList)?.defaultColor).isEqualTo((colorRgb))
```
What ended up being the issue @Akshatkamboj14?

Akshatkamboj14 replied:
actually @BenHenning, I was calling setSupportButtonTintList in the binding adapter, and in the test I was calling getButtonTintList so I think that was the issue, when I changed this (buttonTintList to supportButtonTintList) its calling the get method of SupportButtonTintList and the test is passing.

Akshatkamboj14 replied:
and moreover ben, there's is some error in binding adapter and this error is the reason of the failing Bazel test, can you help me in this?

BenHenning replied:
Left one follow-up comment--PTAL.

Akshatkamboj14 replied:
Okay, Ben.

BenHenning reviewed (changes requested):
@Akshatkamboj14 I think CI is failing--PTAL and fix it, then send the PR back for review.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/BUILD.bazel:
```diff
@@ -27,6 +27,18 @@ _TEST_FILES = [
+    cat $(SRCS) |
+    sed 's/import org.oppia.android.R/import org.oppia.android.app.databinding.R/g' |
+    sed 's/import org.oppia.android.databinding./import org.oppia.android.app.databinding.databinding./g' |
+    sed 's/import org.oppia.android.app.databinding.AppCompatCheckBoxBindingAdapters./import org.oppia.android.app.databinding.AppCompatCheckBoxBindingAdapters_updated./g' > $(OUTS)
```
@Akshatkamboj14 this part is where the issue is coming from. I don't think we need to do the replacement for AppCompatCheckBoxBindingAdapters because it doesn't use any resources (so it isn't being regenerated).

Akshatkamboj14 replied:
so @BenHenning How can we resolve this? should I remove line 38?

BenHenning replied:
Yep.

Akshatkamboj14 committed:
corrected ci check

BenHenning commented:
FYI I can't review this in detail atm because GitHub is currently having a PR outage, but I'll follow up once the PR is sent back to me (since CI still needs to be fixed, anyway).

Akshatkamboj14 commented:
> FYI I can't review this in detail atm because GitHub is currently having a PR outage, but I'll follow up once the PR is sent back to me (since CI still needs to be fixed, anyway).

Okay Sure @BenHenning, I have deleted the line and pushed the changes and assign it back to you to review, since CI checks is not working now you can review it when it get fixed, Thanks.

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionContentViewModel.kt:
```diff
@@ -9,7 +9,8 @@ class SelectionInteractionContentViewModel(
   private val itemIndex: Int,
-  private val selectionInteractionViewModel: SelectionInteractionViewModel
+  private val selectionInteractionViewModel: SelectionInteractionViewModel,
+  val enabledItem: ObservableBoolean
```
Perhaps instead: ``isEnabled`` would be clearer?

Akshatkamboj14 replied:
Okay ben

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdapters.java:
```diff
@@ -0,0 +1,18 @@
+import androidx.appcompat.widget.AppCompatCheckBox;
+import androidx.databinding.BindingAdapter;
+
+/**
+ * Custom data-binding adapters for {@link AppCompatCheckBox}s.
+ */
```
```suggestion
/** Custom data-binding adapters for {@link AppCompatCheckBox}s. */
```

Akshatkamboj14 replied:
Okay ben

Akshatkamboj14 replied:
![image](https://user-images.githubusercontent.com/92685493/221245578-8c4a3e26-467e-437c-bca2-1ff8e1be749f.png)
@BenHenning It's showing this after converting it to a single line.

BenHenning replied:
Ah, I forgot that checkstyle doesn't allow for this (valid per the Google Java style guide which we follow) version--thanks for checking!

Akshatkamboj14 replied:
Thanks ben.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -52,10 +56,17 @@ class SelectionInteractionViewModel private constructor(
       ?: minAllowableSelectionCount
   }
   private val selectedItems: MutableList<Int> = mutableListOf()
+  val enabledItemsList by lazy {
```
Does this need to be public? It seems to only be used within this model unless I'm missing something.

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -584,4 +584,7 @@
   <string name="administrator_controls_fragment_test_activity_label">Administrator Controls Fragment Test Activity</string>
   <string name="revision_navigation_cards_header">Continue Studying</string>
   <string name="state_fragment_hint_bar_forced_announcement_text">Go to the bottom of the screen for a hint.</string>
+  <string name="please_select_all_correct_choices">Please select all correct choices.</string>
```
The names need to make sense in the context of the entire app, and explain what the label is describing not the text of the label itself. For example, this one could be: ``state_fragment_item_selection_no_items_selected_hint_text``. Ditto for the other strings--please update them.

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+import org.robolectric.annotation.LooperMode
+import javax.inject.Singleton
+
+/** Test for [AppCompatCheckBoxBindingAdapters] */
```
```suggestion
/** Test for [AppCompatCheckBox] binding adapters. */
```

``AppCompatCheckBoxBindingAdapters`` isn't itself a class.

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+  application = AppCompatCheckBoxBindingAdaptersTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+
```
```suggestion
```

For consistency.

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+    Intents.release()
+  }
+
+  private var colorRgb: Int = Color.valueOf(-0x10000).toArgb()
```
Nit: please don't mix methods & variables (for consistency with other classes). Should be moved above ``setUp``.

Also, a negative color doesn't really make sense. Why can't this be a positive value?

Akshatkamboj14 replied:
Ben, I have written this here because in the DrawableBindingAdaptersTest this line below the setUp and tearDown https://github.com/oppia/oppia-android/blob/c7349c3c24f99b16c9f5115fd410fbe9c827e1d0/app/src/sharedTest/java/org/oppia/android/app/databinding/DrawableBindingAdaptersTest.kt#L126

If it should be above the setUp then I can do that as well, Thanks.

BenHenning replied:
Ah, that's not right in the existing test, either. Thanks for fixing it here!

Akshatkamboj14 replied:
Thanks.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  private fun setUpTestApplicationComponent() {
+    ApplicationProvider.getApplicationContext<TestApplication>()
+      .inject(this)
```
```suggestion
    ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
```

Nit: can this fit on one line?

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/AppCompatCheckBoxBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,202 @@
+    activityRule.scenario.onActivity {
+      val appCompatCheckBox: AppCompatCheckBox = getAppCompatCheckBox(it)
+      setButtonTint(appCompatCheckBox, colorRgb)
+      assertThat((appCompatCheckBox.supportButtonTintList)?.defaultColor).isEqualTo((colorRgb))
```
```suggestion
      assertThat(appCompatCheckBox.supportButtonTintList?.defaultColor).isEqualTo(colorRgb)
```

Extra parentheses aren't needed.

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -3619,6 +3619,118 @@ class StateFragmentTest {
+      verifyViewTypeIsPresent(SELECTION_INTERACTION)
+      selectItemSelectionCheckbox(optionPosition = 0, expectedOptionText = "Red")
+
+      onView(
+        withId(
+          R.id.selection_interaction_textview
+        )
+      ).check(matches(withText("You may select more choices.")))
```
```suggestion
      onView(withId(R.id.selection_interaction_textview))
        .check(matches(withText("You may select more choices.")))
```

Cleaner and more compact way to wrap. Ditto elsewhere (where applicable).

Akshatkamboj14 replied:
okay

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -3619,6 +3619,118 @@ class StateFragmentTest {
+  }
+
+  @Test
+  fun testStateFragment_interactions_maxRadioItemSelected_selectionTextIsDisplayed() {
```
We should also test that:
- Deselecting from max returns to the "You may select more choices" text
- Deselecting back to 0 returns to the "Please select all correct choices" text
- Not selecting the max number of radio elements should keep the other checkboxes enabled
- Selecting max, then one below, should lead to no other radio elements being enabled

Akshatkamboj14 replied:
okay

BenHenning reviewed (changes requested):
Thanks @Akshatkamboj14. The PR is looking quite good. I mostly had nits and some suggestions for more tests. PTAL.

Akshatkamboj14 committed:
added requested changes

Akshatkamboj14 committed:
solved k-lint issues

Akshatkamboj14 committed:
solved k-lint issues

Akshatkamboj14 committed:
added requested changes

Akshatkamboj14 committed:
solved k-lint issues

Akshatkamboj14 committed:
fixes issue

Akshatkamboj14 committed:
uncommented line

Akshatkamboj14 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into ItemSelectionInteraction-text-change-and-checkboxes-should-be-disabled-on-selection

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -238,7 +238,7 @@ class StateFragmentTest {
     Intents.init()
     setUpTestApplicationComponent()
-    testCoroutineDispatchers.registerIdlingResource()
+//    testCoroutineDispatchers.registerIdlingResource()
```
Did you mean to commit this? We shouldn't check in commented out code, so please revert.

Akshatkamboj14 replied:
I totally forgot about this comment, apologies for this.

BenHenning replied:
No worries.

BenHenning reviewed (commented):
Thanks @Akshatkamboj14! The PR pretty much LGTM, just had one nit.

Besides that, I think you have one comment open from yourself that will need to be closed if it's no longer valid, and I think the PR needs to be updated with the latest develop. Please assign back once those are done, the last comment is addressed, and CI is passing (since then the PR should be ready to merge).

BenHenning reviewed (approved):
Thanks @Akshatkamboj14. This LGTM!

BenHenning commented:
I believe this needs your review as well @rt4914. PTAL.

BenHenning commented:
Also @Akshatkamboj14 please update your PR description to include a proper explanation of what this PR is doing, and to include screenshots and/or videos per the 'UI-specific only PRs' section to show the changed behaviors for both checkbox enabled/disabled states and the new item selection label text.

Akshatkamboj14 commented:
> Also @Akshatkamboj14 please update your PR description to include a proper explanation of what this PR is doing, and to include screenshots and/or videos per the 'UI-specific only PRs' section to show the changed behaviors for both checkbox enabled/disabled states and the new item selection label text.

Okay sure @BenHenning, I'll update the PR description.

Akshatkamboj14 commented:
> Also @Akshatkamboj14 please update your PR description to include a proper explanation of what this PR is doing, and to include screenshots and/or videos per the 'UI-specific only PRs' section to show the changed behaviors for both checkbox enabled/disabled states and the new item selection label text.



> Also @Akshatkamboj14 please update your PR description to include a proper explanation of what this PR is doing, and to include screenshots and/or videos per the 'UI-specific only PRs' section to show the changed behaviors for both checkbox enabled/disabled states and the new item selection label text.

Hey @BenHenning I have updated the PR Description and added supporting screen recording, Thanks.

Akshatkamboj14 commented:
Hey @rt4914 can you PTAL, its needs your review as well to get merged?

rt4914 reviewed (approved):
LGTM
